### PR TITLE
Backend state changes

### PR DESCRIPTION
**Pull request by JoeyDarbyshire**
I took the socket and room objects out of the backend socket.js and moved them to the backend state and implemented immutable.js on them.

### CODE REVIEW

File: server/socket.js [hunk 1]
```diff
@@ -1,18 +1,20 @@
 const chalk = require('chalk');
 const { Map } = require('immutable');
 const store = require('./redux/store');
+
 const { createAndEmitUser, updateUserData, removeUserAndEmit } = require('./redux/reducers/user-reducer');
```
Comment (parky22) [inline server/socket.js]: extra space

File: server/socket.js [hunk 2]
```diff
@@ -1,18 +1,20 @@
 const chalk = require('chalk');
 const { Map } = require('immutable');
 const store = require('./redux/store');
+
```
Comment (parky22) [inline server/socket.js]: extra space before const on line 5

Comment (Jmikeydarby) [inline server/socket.js]: separating the reducers from everything else to make it look neater

Comment (Jmikeydarby) [inline server/socket.js]: I can remove if people don't like that

File: server/socket.js [hunk 3]
```diff
@@ -96,9 +100,9 @@ module.exports = io => {
       const peerId = config.peer_id;
       const iceCandidate = config.ice_candidate;
       console.log(`[${socket.id}] relaying ICE candidate to [${peerId}] ${iceCandidate}`);
-
-      if (peerId in sockets) {
-        sockets[peerId].emit('iceCandidate', { 'peer_id': socket.id, 'ice_candidate': iceCandidate });
+      const sockets = store.getState().sockets;
```
Comment (parky22) [inline server/socket.js:104]: why is this not colored like the other variables? like line 114. might not be a big deal but it's bothering me.

Comment (JoeyDarbyshire) [reply server/socket.js:104]: because it's highlighted as the start of the diff

File: browser/webRTC/client.js [hunk 4]
```diff
@@ -1,7 +1,22 @@
...
 const ICE_SERVERS = [
-  { url: 'stun:stun.l.google.com:19302' }
+  { url: 'stun:stun.l.google.com:19302' },
+  {
+    url: 'turn:numb.viagenie.ca',
+    credential: 'muazkh',
+    username: 'webrtc@live.com'
+  },
+  {
+    url: 'turn:192.158.29.39:3478?transport=udp',
```
Comment (bushidocodes) [inline browser/webRTC/client.js:11]: pinged all of these just to sanity check that they're live, and they responded as expected.

File: server/redux/reducers/room-reducer.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+/* --------------- INITIAL STATE --------------- */
+
+const initialState = Map({});
+
+/* --------------- ACTIONS --------------- */
+
+const ADD_ROOM = 'ADD_ROOM';
+const ADD_SOCKET_TO_ROOM = 'ADD_SOCKET_TO_ROOM';
+const REMOVE_SOCKET_TO_ROOM = 'REMOVE_SOCKET_TO_ROOM';
```
Comment (bushidocodes) [inline server/redux/reducers/room-reducer.js]: sort of semantically confusing wording. I see you're using removeSocketFromRoom below. Can you change this to match?

File: server/redux/reducers/room-reducer.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+/* --------------- REDUCER --------------- */
+
+const roomReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case ADD_ROOM:
+      return state.set(action.room, Map({}));
+    case ADD_SOCKET_TO_ROOM:
+      return state.setIn([action.room, action.socket.id], action.socket);
+    case REMOVE_SOCKET_TO_ROOM:
```
Comment (bushidocodes) [inline server/redux/reducers/room-reducer.js]: save as above

File: browser/webRTC/client.js [hunk 7]
```diff
@@ -1,7 +1,22 @@
 /** You should probably use a different stun server doing commercial stuff **/
 /** Also see: https://gist.github.com/zziuni/3741933 **/
 const ICE_SERVERS = [
-  { url: 'stun:stun.l.google.com:19302' }
```
Comment (bushidocodes) [inline browser/webRTC/client.js:4]: Joey's the #1 Stunna. Wha Wha Wha Wha? https://youtu.be/f56A7obPbtE?t=2m15s

File: server/redux/reducers/room-reducer.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  return {
+    type: ADD_SOCKET_TO_ROOM,
+    room,
+    socket
+  };
+};
+
+const removeSocketFromRoom = (room, socket) => {
+  return {
+    type: REMOVE_SOCKET_TO_ROOM,
```
Comment (bushidocodes) [inline server/redux/reducers/room-reducer.js]: Same as above

**Review by bushidocodes — CHANGES_REQUESTED** [review]
Bosstastic code. You were like the borg. You descended upon the state hiding throughout the server and assimilated into the great collective of Redux. The only teeny change I request is the semantic change to the action to used `REMOVE_SOCKET_FROM_ROOM`. Thank you for being a great American!

File: server/socket.js (hunk 3 above)
Comment (bethqiang) [inline server/socket.js]: ^lol

File: server/socket.js [hunk 9]
```diff
@@ -54,21 +55,22 @@ module.exports = io => {
...
-      }
-      for (const id in rooms[room]) {
-        rooms[room][id].emit('addPeer', { 'peer_id': socket.id, 'should_create_offer': false });
-        socket.emit('addPeer', { 'peer_id': id, 'should_create_offer': true });
+      if (!(store.getState().rooms.has(room))) {
+        console.log(`Adding ${room} to state`);
+        store.dispatch(addRoom(room));
       }
-      rooms[room][socket.id] = socket;
+      store.getState().rooms.get(room).valueSeq().forEach(peer => {
```
Comment (bethqiang) [inline server/socket.js]: To make this (& line 86) slightly more consumable (???), I'd maybe save part of it as a variable, but I also understand that that variable wouldn't be used anywhere else sooo...

...idk I'm conflicted. Not really asking you to change anything. Just expressing the feelz.

**Review by bethqiang — APPROVED** [review]

File: server/socket.js (hunk 3 above)
Comment (bethqiang) [reply server/socket.js:104]: ^LOL

**Pull request merged**